tests/services/test_planning_svc.py

```diff
@@ -352,3 +361,109 @@ def test_duplicate_lateral_filter(self, loop, setup_planning_test, planning_svc,
+
+        input_facts = [
+            Fact(trait='a', value='1'),
+            Fact(trait='a.b', value='2'),
+            Fact(trait='a.b.c', value='3'),
+            Fact(trait='server', value='5')
+        ]
+
+        new_links = loop.run_until_complete(planning_svc.add_test_variants([link], agent, facts=input_facts))
+        assert len(new_links) == 2
+
+        found_commands = set(x.command for x in new_links)
+        assert len(found_commands) == 2  # the original and the replaced
+        assert encoded_command in found_commands
+        assert BaseWorld.encode_string('2') in found_commands
+
+    def test_trait_with_three_parts(self, loop, setup_planning_test, planning_svc):
+        _, agent, operation, ability = setup_planning_test
+        encoded_command = BaseWorld.encode_string('#{a.b.c}')
+        link = Link.load(dict(command=encoded_command, paw=agent.paw, ability=ability, status=0))
+
+        input_facts = [
+            Fact(trait='a', value='1'),
+            Fact(trait='a.b', value='2'),
+            Fact(trait='a.b.c', value='3'),
+            Fact(trait='server', value='5')
+        ]
+
+        new_links = loop.run_until_complete(planning_svc.add_test_variants([link], agent, facts=input_facts))
+        assert len(new_links) == 2
+
+        found_commands = set(x.command for x in new_links)
+        assert len(found_commands) == 2  # the original and the replaced
+        assert encoded_command in found_commands
+        assert BaseWorld.encode_string('3') in found_commands
+
+    def test_trait_with_multiple_variations_of_parts(self, loop, setup_planning_test, planning_svc):
+        _, agent, operation, ability = setup_planning_test
+        encoded_command = BaseWorld.encode_string('#{a} #{a.b} #{a.b.c}')
+        link = Link.load(dict(command=encoded_command, paw=agent.paw, ability=ability, status=0))
+
+        input_facts = [
+            Fact(trait='a', value='1'),
+            Fact(trait='a.b', value='2'),
+            Fact(trait='a.b.c', value='3'),
+            Fact(trait='server', value='5')
+        ]
+
+        new_links = loop.run_until_complete(planning_svc.add_test_variants([link], agent, facts=input_facts))
+        assert len(new_links) == 2
+
+        found_commands = set(x.command for x in new_links)
+        assert len(found_commands) == 2  # the original and the replaced
+        assert encoded_command in found_commands
+        assert BaseWorld.encode_string('1 2 3') in found_commands
+
+    async def test_remove_links_missing_facts_keeps_link_without_facts(self, ability):
+        cmd = 'a -b --foo={bar}'  # almost includes a fact, but missing a '#' in front of '{bar}'
+        links = [Link(command=BaseWorld.encode_string(cmd), paw='1', ability=ability())]
+        await BasePlanningService.remove_links_missing_facts(links)
+        assert len(links) == 1
+        assert links[0].raw_command == cmd
+
+    async def test_remove_links_missing_facts_removes_part_part_fact(self, ability):
```

Comment:
Should this be `test_remove_links_missing_facts_removes_one_part_fact`?

Reply:
I don't know what's confusing about `part_part` here :) ..Yeaaaa I'll fix that.